pkg/api/types.go

```diff
@@ -320,8 +320,64 @@ type Result struct {
 
 	// ImageID describes resulting image ID.
 	ImageID string
+
+	// BuildInfo holds information about the result of a build.
+	BuildInfo []BuildStepInfo
+}
+
+// BuildStepInfo holds information about a particular step in the build process.
+type BuildStepInfo struct {
+	// FailureReason is a camel case reason that is used by the machine to reply
+	// back to the OpenShift builder with information why any of the steps in the
+	// build, failed.
+	FailureReason StepFailureReason
 }
 
+// StepFailureReason holds the type of failure that occured during the build
+// process.
+type StepFailureReason string
+
+const (
+	// ReasonAssembleFailed is the reason associated with the Assemble script
+	// failing.
+	ReasonAssembleFailed StepFailureReason = "AssembleFailed"
+
+	// ReasonRuntimeAssembleFailed is the reason associated with the Assemble
+	// Runtime failing.
+	ReasonRuntimeAssembleFailed StepFailureReason = "RuntimeAssembleFailed"
+
+	// ReasonPullBuilderImageFailed is the reason associated with failing to pull
+	// the builder image.
+	ReasonPullBuilderImageFailed StepFailureReason = "PullBuilderImageFailed"
+
+	// ReasonPullRuntimeImageFailed is the reason associated with failing to pull
+	// the runtime image.
+	ReasonPullRuntimeImageFailed StepFailureReason = "PullRuntimeImageFailed"
+
+	// ReasonCommitContainerFailed is the reason associated with failing to
+	// commit the container to the final image.
+	ReasonCommitContainerFailed StepFailureReason = "ContainerCommitFailed"
+
+	// ReasonFetchSourceFailed is the reason associated with failing to download
+	// the source of the build.
+	ReasonFetchSourceFailed StepFailureReason = "FetchSourceFailed"
+
+	// ReasonDockerImageBuildFailed is the reasons associated with a failed
+	// Docker image build.
+	ReasonDockerImageBuildFailed StepFailureReason = "DockerImageBuildFailed"
+
+	// ReasonInvalidArtifactsMappingFailed is the reason associated with an
+	// invalid artifacts mapping of files that need to be copied.
+	ReasonInvalidArtifactsMappingFailed StepFailureReason = "InvalidArtifactsMappingFailed"
+
+	// ReasonFailedFileSystemOperation is the reason associated with a failed fs
+	// operation. create, remove directory, copy file, etc.
+	ReasonFailedFileSystemOperation StepFailureReason = "FileSystemOperationFailed"
+
+	// ReasonInstallScriptsFailed
+	ReasonInstallScriptsFailed StepFailureReason = "InstallScriptsFailed"
```

Comment:
@bparees how far the rabbit hole should i go? trying to account for all failure steps i think is a lot. on the other hand returning only some failure cases to the user in openshift and nothing for others doesn't help either.
i thought about returning more generic failures, but telling the user "build failed at Execute step" doesn't seem i'm helping him identify anything.